src/main/java/org/dependencytrack/persistence/WorkflowStateQueryManager.java

```diff
@@ -0,0 +1,121 @@
+        //update fields
+        WorkflowState workflowState = getObjectById(WorkflowState.class, transientWorkflowState.getId());
+        if (workflowState != null) {
+            workflowState.setStatus(transientWorkflowState.getStatus());
+            workflowState.setUpdatedAt(transientWorkflowState.getUpdatedAt());
+            workflowState.setFailureReason(transientWorkflowState.getFailureReason());
+            return persist(workflowState);
+        }
+        return null;
+    }
+
+    public void deleteWorkflowState(WorkflowState workflowState) {
+        delete(workflowState);
+    }
+
+    public List<WorkflowState> getAllWorkflowStatesForAToken(UUID token) {
+        final Query<WorkflowState> query = pm.newQuery(WorkflowState.class, "this.token == :token");
+        query.setParameters(token);
+        return query.executeList();
+    }
+
+    public List<WorkflowState> getAllWorkflowStatesForParentByToken(UUID token, WorkflowState parentWorkflowState) {
+        StringBuilder filterCriteria = new StringBuilder();
+        filterCriteria = parentWorkflowState == null ? filterCriteria.append(" WHERE PARENT_STEP_ID IS NULL ") :
+                filterCriteria.append(" WHERE PARENT_STEP_ID = ? ");
+        filterCriteria = filterCriteria.append(" AND TOKEN = ? ");
+
+        StringBuilder cteQuery = new StringBuilder();
+        cteQuery = cteQuery.append(PARENT_QUERY).append(filterCriteria).append(UNION_ALL).append(RECURSIVE_QUERY).append(SELECT_QUERY);
+
+        Connection connection = null;
+        PreparedStatement preparedStatement = null;
+        ResultSet rs = null;
+        List<WorkflowState> results = new ArrayList<>();
+        try {
+            connection = (Connection) pm.getDataStoreConnection();
+            if (DbUtil.isMssql() || DbUtil.isOracle()) { // Microsoft SQL Server and Oracle DB already imply the "RECURSIVE" keyword in the "WITH" clause, therefore it is not needed in the query
+                preparedStatement = connection.prepareStatement("WITH " + cteQuery);
+            } else { // Other Databases need the "RECURSIVE" keyword in the "WITH" clause to correctly execute the query
+                preparedStatement = connection.prepareStatement("WITH RECURSIVE " + cteQuery);
+            }
+
+            if (parentWorkflowState == null) {
+                preparedStatement.setObject(1, token);
+            } else {
+                preparedStatement.setObject(1, parentWorkflowState.getId());
+                preparedStatement.setObject(2, parentWorkflowState.getToken());
+            }
+
+            preparedStatement.execute();
+            rs = preparedStatement.getResultSet();
+            while (rs.next()) {
+                WorkflowState workflowState = new WorkflowState();
+                workflowState.setId(rs.getLong(1));
+                WorkflowState parent = new WorkflowState();
+                parent.setId(rs.getLong(2));
+                workflowState.setParent(parent);
```

Comment:
Hmmm... I wonder if this can become confusing when `parent` is only ever populated with an ID. Normally `QueryManager`s return persistent objects that would lazy-load such relationships, but that is not the case here. 

It would be good if this was documented in the javadoc. Alternatively, we could use a dedicated projection class that holds a `parentId` instead of a `parent` object.